### PR TITLE
Feature: `--warn-as-error` flag to make all warnings be errors

### LINKED ISSUE #445: Add a `--fail-on-warning` option or  `--max-warnings=0`
I think there should be a way for a Cadl author to enforce that no warning are left when making a change.

Warning could be nice to ignore while writing the spec but should ultimately be resolved.

Adding this flag would enable enforcing that no warning are left uncheck in the CI.

Alternative?:
- Fail on warning by default and flag to ignore fail?

### PR DESCRIPTION
**Pull request by timotheeguerin**
fix #445

image

### CODE REVIEW

**Comment by nguerrera** [comment]
The paradigm that I'm used to for this is a flag to treat warnings as errors. csc has `/warnaserror,` clan and gcc have `-Werror`.  Usually there is also a mechanism for turning certain warnings into errors rather than all warnings. If we have the primitive concept of upgrading a diagnostic from warning to error then we can have those knobs as well based on that. The concept is effectively the reverse of suppression.

**Comment by nguerrera** [comment]
Also --no-warnings sounds a bit ambiguous, it could also be guessed that this suppresses all warnings.

**Comment by timotheeguerin** [comment]
Does this make the warning show up as error completely or it just treat them like a failing condition?

I guess that make sense with the name was looking for something shorter but `fail-on-warning` would be clearer.

**Comment by timotheeguerin** [comment]
half read your comment, guess you already answered the 1st question.

**Comment by nguerrera** [comment]
Yep, just to be clear, with csc /warnaserror, warnings shows up as an error, which in turn causes the exit code to be non-zero.

**Comment by nguerrera** [comment]
```
$ cat test.cs
class P {
  static void Main() {
    int unused = 0;
  }
}

$ csc test.cs
Microsoft (R) Visual C# Compiler version 4.2.0-3.22181.8 (a59a22c6)
Copyright (C) Microsoft Corporation. All rights reserved.

test.cs(3,9): warning CS0219: The variable 'unused' is assigned but its value is never used

C:\Temp
$ echo $LASTEXITCODE
0

$ csc /warnaserror test.cs
Microsoft (R) Visual C# Compiler version 4.2.0-3.22181.8 (a59a22c6)
Copyright (C) Microsoft Corporation. All rights reserved.

test.cs(3,9): error CS0219: The variable 'unused' is assigned but its value is never used

C:\Temp
$ echo $LASTEXITCODE
1
```

**Comment by nguerrera** [comment]
And clang is the same:

```
$ cat test.c
int main() {
  int unused = 0;
  return 0;
}

$ clang -Wall test.c
test.c:2:7: warning: unused variable 'unused' [-Wunused-variable]
  int unused = 0;
      ^
1 warning generated.

$ echo $?
0

$ clang -Wall -Werror test.c
test.c:2:7: error: unused variable 'unused' [-Werror,-Wunused-variable]
  int unused = 0;
      ^
1 error generated.
```

**Comment by timotheeguerin** [comment]
Does seem to be the pattern followed by most compiler. Rust is the only one I could see that has `--deny warnings` instead.

Don't have a strong opinion on one way or the other. What do you think for the name(We could have alias to ones form other compilers too)
- `--warn-as-error`
- `--warnaserror`

**Comment by nguerrera** [comment]
--warn-as-error works for me

File: packages/compiler/core/cli.ts [hunk 1]
```diff
@@ -93,10 +92,10 @@ async function main() {
             choices: ["error", "warn", "info", "verbose", "debug"],
             describe: "Diagnostics of this level or above will be reported.",
           })
-          .option("no-warnings", {
+          .option("warn-as-error", {
             type: "boolean",
             default: false,
-            describe: "Return non zero exit code if there is any warnings.",
+            describe: "Treat warnings as error and return non-zero exit code if there is any.",
```
Comment (nguerrera) [inline packages/compiler/core/cli.ts:98]:
```suggestion
            describe: "Treat warnings as errors and return non-zero exit code if there are any.",
```

**Review by nguerrera — APPROVED** [review]

**Pull request merged**